File: solutions/c/high-scores/1/high_scores.c
#include "high_scores.h"
/// Return the latest score.
int32_t latest(const int32_t *scores, size_t scores_len){
    return scores[scores_len - 1];
}

/// Return the highest score.
int32_t personal_best(const int32_t *scores, size_t scores_len){
    int mayor = 0;
    for(size_t i = 0; i < scores_len; i++){
        if(scores[i] > mayor) mayor = scores[i];
    }
    return mayor;
}
/// Write the highest scores to `output` (in non-ascending order).
/// Return the number of scores written.
size_t personal_top_three(const int32_t *scores, size_t scores_len, int32_t *output){
    int top1 = 0;
    int top2 = 0;
    int top3 = 0;
    for (size_t i = 0; i < scores_len; i++) {
        if (scores[i] > top1) {
            top3 = top2;
            top2 = top1;
            top1 = scores[i];
        } else if (scores[i] > top2) {
            top3 = top2;
            top2 = scores[i];
        } else if (scores[i] > top3) {
            top3 = scores[i];
        }
    }
    output[0] = top1;
    output[1] = top2;
    output[2] = top3;
    size_t puntajes = 0;
    for(int i = 0; i < 3; i++){
        if(output[i] != 0) puntajes++;
    }
    return puntajes;
}

File: solutions/zig/high-scores/1/high_scores.zig
const std = @import("std");

pub const HighScores = struct {
    scores: []const i32,
    top_three: [3]i32,
    top_three_len: usize,

    pub fn init(scores: []const i32) HighScores {
        const n = @min(3, scores.len);
        var hs = HighScores{
            .scores = scores,
            .top_three = undefined,
            .top_three_len = n,
        };

        for (0..n) |i| hs.top_three[i] = scores[i];
        std.sort.pdq(i32, hs.top_three[0..n], {}, std.sort.desc(i32));

        for (scores[n..]) |score| {
            if (score > hs.top_three[n - 1]) {
                hs.top_three[n - 1] = score;
                std.sort.pdq(i32, hs.top_three[0..n], {}, std.sort.desc(i32));
            }
        }

        return hs;
    }

    pub fn latest(self: *const HighScores) ?i32 {
        const len = self.scores.len;
        return if (len == 0) null else self.scores[len - 1];
    }

    pub fn personalBest(self: *const HighScores) ?i32 {
        const idx = std.sort.argMax(i32, self.scores, {}, std.sort.asc(i32));
        return if (idx == null) null else self.scores[idx.?];
    }

    pub fn personalTopThree(self: *const HighScores) []const i32 {
        return self.top_three[0..self.top_three_len];
    }
};
